From 4e8bbb9109df1dfba371dbc8f7071820fd395e4f Mon Sep 17 00:00:00 2001
From: cheney <11722200@qq.com>
Date: Tue, 22 Jul 2025 16:26:51 +0800
Subject: [PATCH 420/533] Changelogs: 1. Limited the maximum CPU frequency to
 1.4GHz to avoid system crashes at 1.8GHz. 2. Updated the LPCPU module to
 improve overall power efficiency.

---
 .../eswin/eic7702-deepcomputing-fml13v03.dts  |  14 ++-
 .../dts/eswin/eic7702-som560-interleave.dtsi  |   2 +-
 drivers/soc/eswin/eswin-lpcpu.c               | 111 +++++++++++-------
 3 files changed, 78 insertions(+), 49 deletions(-)

diff --git a/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03.dts b/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03.dts
index e90c0f3cbfcc..8bc250425a1f 100644
--- a/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03.dts
+++ b/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03.dts
@@ -307,13 +307,13 @@ &d0_clock {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_die0_gpio94_default>;
 	cpu-voltage-gpios = <&portc 30 GPIO_ACTIVE_HIGH>;
-	cpu-default-frequency = <CLK_FREQ_1800M>;
+	cpu-default-frequency = <CLK_FREQ_1400M>;
 };
 
 &d1_clock {
 	status = "okay";
 	cpu-voltage-gpios = <&portc 30 GPIO_ACTIVE_HIGH>;
-	cpu-default-frequency = <CLK_FREQ_1800M>;
+	cpu-default-frequency = <CLK_FREQ_1400M>;
 };
 
 &d0_reset {
@@ -517,8 +517,10 @@ &d1_ipc_scpu {
 &d0_lpcpu {
 	status = "okay";
 	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_die0_gpio0_default>;
-	irq-gpios = <&porta 0 GPIO_ACTIVE_HIGH>;
+	pinctrl-0 = <&pinctrl_die0_gpio0_default &pinctrl_die1_gpio7_on_default>;
+	irq-gpios = <&porta 0 GPIO_ACTIVE_HIGH>, <&d1_porta 7 IRQ_TYPE_EDGE_BOTH>;
+	soc-vdd-ctrl = <21 0 0>;
+	d2d-power-ctrl = <18 0 0>;
 };
 
 &d1_lpcpu {
@@ -1709,11 +1711,11 @@ &d1_llc_dev{
 };
 
 &d2d {
-	status = "okay";
+	status = "disabled";
 };
 
 &d1_d2d {
-	status = "okay";
+	status = "disabled";
 };
 
 &pinctrl_die1_gpio69_default {
diff --git a/arch/riscv/boot/dts/eswin/eic7702-som560-interleave.dtsi b/arch/riscv/boot/dts/eswin/eic7702-som560-interleave.dtsi
index 72c192ef0bcf..f96e10be2be2 100644
--- a/arch/riscv/boot/dts/eswin/eic7702-som560-interleave.dtsi
+++ b/arch/riscv/boot/dts/eswin/eic7702-som560-interleave.dtsi
@@ -501,7 +501,7 @@ &d0_lpcpu {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_die0_gpio0_default &pinctrl_die1_gpio7_on_default>;
 	irq-gpios = <&porta 0 GPIO_ACTIVE_HIGH>, <&d1_porta 7 IRQ_TYPE_EDGE_BOTH>;
-	vddctrl = <21 0 0>;
+	soc-vdd-ctrl = <21 0 0>;
 };
 
 &d1_lpcpu {
diff --git a/drivers/soc/eswin/eswin-lpcpu.c b/drivers/soc/eswin/eswin-lpcpu.c
index 27c2d0d76cf0..b56b859b1dcd 100644
--- a/drivers/soc/eswin/eswin-lpcpu.c
+++ b/drivers/soc/eswin/eswin-lpcpu.c
@@ -52,7 +52,7 @@ struct lowpower_info {
 };
 
 #define LPCPU_IOC_MAGIC 'L'
-#define LPCPU_IOC_SAVEPOWER                          \
+#define LPCPU_IOC_SAVEPOWER						  \
 	_IOWR(LPCPU_IOC_MAGIC, 0x1, struct lowpower_info)
 
 #define FW_LOAD_UNKNOW 0
@@ -148,15 +148,15 @@ static int eswin_lpcpu_open(struct inode *inode, struct file *filp)
 }
 
 static ssize_t eswin_lpcpu_read(struct file *filp, char __user *userbuf,
-				      size_t count, loff_t *ppos)
+					  size_t count, loff_t *ppos)
 {
 	printk("lpcpu read\n");
 	return 0;
 }
 
 static ssize_t eswin_lpcpu_write(struct file *filp,
-				       const char __user *userbuf, size_t count,
-				       loff_t *ppos)
+					   const char __user *userbuf, size_t count,
+					   loff_t *ppos)
 {
 	printk("lpcpu write\n");
 	u8 msg[8];
@@ -192,7 +192,7 @@ static int eswin_lpcpu_release(struct inode *inode, struct file *filp)
 }
 
 static long eswin_lpcpu_ioctl(struct file *filp, unsigned int cmd,
-			    unsigned long arg)
+				unsigned long arg)
 {
 	struct lowpower_info lowpower;
 	struct device *dev = lpcpu->mdev.parent;
@@ -286,16 +286,18 @@ static const struct file_operations eswin_lpcpu_ops = {
 	.unlocked_ioctl = eswin_lpcpu_ioctl,
 };
 
-static int lpcpu_boot_status(struct mbox_chan *mbox_channel, u32 vdd_config_msg)
+static int lpcpu_boot_status(struct mbox_chan *mbox_channel, u32 boot_config_msg)
 {
 	int ret = 0;
 	u8 msg[8];
 	msg[0] = 0xca;
 	msg[1] = 0xec;
 	msg[2] = 0x55;
-    // vdd config msg
-    msg[4] = vdd_config_msg & 0xff;
-    msg[7] = (vdd_config_msg >> 24) & 0xff;
+	// boot config msg
+	msg[4] = boot_config_msg & 0xff;
+	msg[5] = (boot_config_msg >> 8) & 0xff;
+	msg[6] = (boot_config_msg >> 16) & 0xff;
+	msg[7] = (boot_config_msg >> 24) & 0xff;
 
 	ret = mbox_send_message(mbox_channel, msg);
 	if (ret < 0){
@@ -316,8 +318,9 @@ static int eswin_lpcpu_probe(struct platform_device *pdev)
 	int numa_id = 0;
 	int ret;
 	long timeout;
-    u32 vdd_config_gpio = 0, vdd_config_die = 0, vdd_config_polarity = 0;
-    u32 buff[3], vdd_config_msg = 0;
+	u32 vdd_config_gpio = 0, vdd_config_die = 0, vdd_config_polarity = 0;
+	u32 d2d_power_config_gpio = 0, d2d_power_config_die = 0, d2d_power_config_polarity = 0;
+	u32 buff[3], boot_config_msg = 0;
 
 	lpcpu = devm_kzalloc(dev, sizeof(*lpcpu), GFP_KERNEL);
 	if (!lpcpu)
@@ -358,32 +361,56 @@ static int eswin_lpcpu_probe(struct platform_device *pdev)
 		goto err_mailbox;
 	}
 
-    /* soc vdd config
-     * vdd_config_msg:  31      can not be used
-     *                  30      vdd config exists 1: exist 0: not exist
-     *                  29:25   reserved
-     *                  24      soc vdd polarity 0: positive 1: negative
-     *                  23:8    reserved
-     *                  7       die number 0-1, ignored on single die chips
-     *                  6:0     gpio number 0-127
-     * of_property:     vddctrl = <GPIO [POLARITY] [DIE]>;
-     *                  GPIO:       gpio port number which vdd control uses, range: 0-127
-     *                  POLARITY:   gpio level polarity, default: 0
-     *                              0: positive, high -> high vdd, low -> low vdd
-     *                              1: negative, high -> low vdd, low -> high vdd
-     *                  DIE:        die number, range 0-1, default: 0 */
-    ret = of_property_read_variable_u32_array(pdev->dev.of_node, "vddctrl", buff, 1, 3);
-    if (ret > 0) {
-        vdd_config_gpio = buff[0] & 0x7f;
-        if (ret > 1)
-            vdd_config_polarity = buff[1] ? 1 : 0;
-        if (ret > 2)
-            vdd_config_die = buff[2] ? 1 : 0;
-        vdd_config_msg = vdd_config_gpio | (vdd_config_die << 7) |
-            (vdd_config_polarity << 24) | (1 << 30);
-        dev_info(dev, "vdd config gpio: die %d, port %d, polarity %s\n", vdd_config_die,
-                vdd_config_gpio, !vdd_config_polarity ? "positive" : "negative");
-    }
+	/* Get boot config
+	 * boot_config_msg: 31      can not be used
+	 *                  30      soc vdd config exists 1: exist 0: not exist
+	 *                  29      d2d power config exists 1: exist 0: not exist
+	 *                  28:26   reserved
+	 *                  25      d2d power polarity 0: positive 1: negative
+	 *                  24      soc vdd polarity 0: positive 1: negative
+	 *                  23:8    reserved
+	 *                  15      d2d power die number 0-1, ignored on single die chip
+	 *                  14:8    d2d power gpio number 0-127
+	 *                  7       soc vdd die number 0-1, ignored on single die chips
+	 *                  6:0     soc vdd gpio number 0-126, 127 for invalid
+	 * of_property:     soc-vdd-ctrl = <GPIO [POLARITY] [DIE]>;
+	 *                  GPIO:       gpio port number which vdd control uses,
+	 *                              range: 0-126, 127 for invalid
+	 *                  POLARITY:   gpio level polarity, default: 0
+	 *                              0: positive, high -> high vdd, low -> low vdd
+	 *                              1: negative, high -> low vdd, low -> high vdd
+	 *                  DIE:        die number, range 0-1, default: 0
+	 *                  d2d-power-ctrl = <GPIO [POLARITY] [DIE]>;
+	 *                  GPIO:       gpio port number which d2d power control uses,
+	 *                              range: 0-126, 127 for invalid
+	 *                  POLARITY:   gpio level polarity, default: 0
+	 *                              0: positive, high -> power on, low -> power off
+	 *                              1: negative, high -> power off, low -> power on
+	 *                  DIE:        die number, range 0-1, default: 0 */
+	ret = of_property_read_variable_u32_array(pdev->dev.of_node, "soc-vdd-ctrl", buff, 1, 3);
+	if (ret > 0) {
+		vdd_config_gpio = buff[0] & 0x7f;
+		if (ret > 1)
+			vdd_config_polarity = buff[1] ? 1 : 0;
+		if (ret > 2)
+			vdd_config_die = buff[2] ? 1 : 0;
+		boot_config_msg = vdd_config_gpio | (vdd_config_die << 7) |
+			(vdd_config_polarity << 24) | (1 << 30);
+		dev_info(dev, "soc vdd config gpio: die %d, port %d, polarity %s\n", vdd_config_die,
+				vdd_config_gpio, !vdd_config_polarity ? "positive" : "negative");
+	}
+	ret = of_property_read_variable_u32_array(pdev->dev.of_node, "d2d-power-ctrl", buff, 1, 3);
+	if (ret > 0) {
+		d2d_power_config_gpio = buff[0] & 0x7f;
+		if (ret > 1)
+			d2d_power_config_polarity = buff[1] ? 1 : 0;
+		if (ret > 2)
+			d2d_power_config_die = buff[2] ? 1 : 0;
+		boot_config_msg |= (d2d_power_config_gpio << 8) | (d2d_power_config_die << 15) |
+			(d2d_power_config_polarity << 25) | (1 << 29);
+		dev_info(dev, "d2d power config gpio: die %d, port %d, polarity %s\n", vdd_config_die,
+				d2d_power_config_gpio, !d2d_power_config_polarity ? "positive" : "negative");
+	}
 
 	mutex_init(&lpcpu->lock);
 	init_waitqueue_head(&lpcpu->waitq);
@@ -429,7 +456,7 @@ static int eswin_lpcpu_probe(struct platform_device *pdev)
 		goto err_mmio;
 	}
 
-	ret = lpcpu_boot_status(lpcpu->mbox_channel, vdd_config_msg);
+	ret = lpcpu_boot_status(lpcpu->mbox_channel, boot_config_msg);
 	if (ret < 0) {
 		dev_err(dev, "Send message to lpcpu via mailbox failed!\n");
 		goto err_mmio;
@@ -480,12 +507,12 @@ static const struct of_device_id eswin_lpcpu_match[] = {
 };
 
 static struct platform_driver eswin_lpcpu_driver = {
-    .driver = {
+	.driver = {
 	.name = "win2030-lpcpu",
 	.of_match_table = eswin_lpcpu_match,
-    },
-    .probe = eswin_lpcpu_probe,
-    .remove = eswin_lpcpu_remove,
+	},
+	.probe = eswin_lpcpu_probe,
+	.remove = eswin_lpcpu_remove,
 };
 
 static int __init lpcpu_modules_init(void)
-- 
2.52.0

